src/services/hana.py

```diff
@@ -0,0 +1,55 @@
+import logging
+from collections.abc import Generator
+
+from hdbcli import dbapi
+
+from utils.settings import (
+    DATABASE_PASSWORD,
+    DATABASE_PORT,
+    DATABASE_URL,
+    DATABASE_USER,
+)
+from utils.singleton_meta import SingletonMeta
+
+
+class Hana(metaclass=SingletonMeta):
+    """
+    Manages a singleton connection to the Hana database.
+
+    This class establishes a connection to the Hana database using the provided
+    credentials and configuration. It ensures that only one instance of the connection
+    exists at any time by utilizing the SingletonMeta metaclass.
+    """
+
+    connection: dbapi.Connection | None = None
+
+    def set_connection(self, connection: dbapi.Connection) -> None:
+        """
+        sets the database connection for the service.
+
+        args:
+            connection (dbapi.connection): the database connection object.
+        """
+        self.connection = connection
+
+
+def get_hana_connection() -> Generator[Hana, None, None]:
+    """Create a connection to the Hana database."""
+    try:
+        hana = Hana()
+        if not hana.connection:
+            hana.set_connection(
```

Comment:
This could have been done in the constructor of `Hana` class. Please check the purpose of `metaclass=SingletonMeta`.

Also an example: https://www.perplexity.ai/search/give-an-example-of-singleton-c-g8sTEUiaTnSVoTRB.Y9YSw#0

Reply:
Thanks for sharing a link to a perplexity search result. I guess.

Reply:
```
def __init__(self, connection | None = None):
   self._connection = connection or some_logic_to_create_a_new_collection()
```

Reply:
https://www.perplexity.ai/search/-3GT0269HQdmyK7i9ssTdWQ